Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,185 @@
+  queue.Enqueue([&] {  // clang-format off
+    // clang-format on
+    EXPECT_ANY_THROW(queue.Enqueue([] {}););
+    signal_finished();
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+}
+
+TEST_P(AsyncQueueTest, EnqueueAllowingNestingWorksFromWithinEnqueue) {
+  queue.Enqueue([&] {  // clang-format off
+    queue.EnqueueAllowingNesting([&] { signal_finished(); });
+    // clang-format on
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+}
+
+TEST_P(AsyncQueueTest, EnqueueBlocking) {
+  bool finished = false;
+  queue.EnqueueBlocking([&] { finished = true; });
+  EXPECT_TRUE(finished);
+}
+
+TEST_P(AsyncQueueTest, EnqueueBlockingDisallowsNesting) {
+  queue.EnqueueBlocking([&] {  // clang-format off
+    EXPECT_ANY_THROW(queue.EnqueueBlocking([] {}););
+    // clang-format on
+  });
+}
+
+TEST_P(AsyncQueueTest, StartExecutionDisallowsNesting) {
+  queue.EnqueueBlocking(
+      [&] { EXPECT_ANY_THROW(queue.StartExecution([] {});); });
+}
+
+TEST_P(AsyncQueueTest, VerifyCalledFromOperationWorksWithOperationInProgress) {
+  queue.EnqueueBlocking(
+      [&] { EXPECT_NO_THROW(queue.VerifyCalledFromOperation()); });
+}
+
+TEST_P(AsyncQueueTest, CanScheduleOperationsInTheFuture) {
+  std::string steps;
+
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.Enqueue([&] {
+    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), kTimerId1, [&] {
+      steps += '4';
+      signal_finished();
+    });
+    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(1), kTimerId2,
+                            [&steps] { steps += '3'; });
+    queue.EnqueueAllowingNesting([&steps] { steps += '2'; });
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+  EXPECT_EQ(steps, "1234");
+}
+
+TEST_P(AsyncQueueTest, CanCancelDelayedCallbacks) {
+  std::string steps;
+
+  queue.Enqueue([&] {
+    // Queue everything from the queue to ensure nothing completes before we
+    // cancel.
+
+    queue.EnqueueAllowingNesting([&steps] { steps += '1'; });
+
+    DelayedOperation delayed_operation = queue.EnqueueAfterDelay(
+        AsyncQueue::Milliseconds(1), kTimerId1, [&steps] { steps += '2'; });
+
+    queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), kTimerId2, [&] {
+      steps += '3';
+      signal_finished();
+    });
+
+    EXPECT_TRUE(queue.IsScheduled(kTimerId1));
+    delayed_operation.Cancel();
+    EXPECT_FALSE(queue.IsScheduled(kTimerId1));
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+  EXPECT_EQ(steps, "13");
+  queue.EnqueueBlocking([&] { EXPECT_FALSE(queue.IsScheduled(kTimerId1)); });
+}
+
+TEST_P(AsyncQueueTest, DelayedOperationIsValidAfterTheOperationHasRun) {
+  DelayedOperation delayed_operation;
+  queue.Enqueue([&] {
+    delayed_operation = queue.EnqueueAfterDelay(
+        AsyncQueue::Milliseconds(10), kTimerId1, [&] { signal_finished(); });
+    EXPECT_TRUE(queue.IsScheduled(kTimerId1));
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish());
+  queue.EnqueueBlocking([&] { EXPECT_FALSE(queue.IsScheduled(kTimerId1)); });
```

Comment:
I am confused. There is already `EXPECT_TRUE(WaitForTestToFinish());`. Is it neccessary to put the checking in the queue instead of doing it directly here?